### PR TITLE
Registry Module

### PR DESCRIPTION
**Pull request by camsoftworks2018**
Feel free to deny the PR if the module should be updated to fit jake's logging module and maybe a future 'per module' virtual filesystem module

### CODE REVIEW

**Comment by camsoftworks2018** [comment]
tests run fine,
Give the code a look and absolutely give me pointers on how to do better.

Marked draft as i think ill utilize jake's logging module instead of panicking when filesystem funkies happen

**Comment by mavic7** [comment]
Please bear with while I inspect this module and let you know whether it'll work with the logging. I have only implemented local module name translation in the logging module for testing so I expect to be flexible with registry

File: Cargo.toml
Comment (jw2476) [inline Cargo.toml]: could the crate be renamed to `birb_registry` otherwise it will conflict on crates.io when we come to publish

File: birb_registry/Cargo.toml
Comment (jw2476) [inline birb_registry/Cargo.toml]: you'll also need to rename it here

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: in my opinion, the Registry shouldn't be responsible for keeping track of time, or autosave intervals either, that should be up to the user, just provide a save function that takes a path and dumps to a file and maybe provide a timed autosave module as well that stores an interval. Maybe wait until the clock module gets merged

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: You can use Self::default() here instead of Registry::default(), clippy will complain about this

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: you can reduce this to a one line implicit return

```rust
serde_json::from_slice(self.data.get(&key).unwrap()).unwrap()
```

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: This can be derived now since you aren't using sqlite anymore

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: I'd argue this should be a Module instead of a MainThreadModule, I don't see any reason it should be single-threaded

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: don't swear in code

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: if you use this on a Path you can avoid the unstable feature

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: If you want to panic on an error, you could use unwrap

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: use lower case for variables otherwise this will get confused with types

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: This should be put in a load function rather than being based on some timing system

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: same here, use a lower case `E` for variables

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: unwrap instead of manual panic

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: unwrap instead of manual panic

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: unwrap instead of manual panic

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: unwrap instead of manual panic

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: make this into a save function rather than tick-based

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: this will cause registry.json to be read back in after a write, possibly causing overwrites and bad for performance

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs]: this anchors loading and saving to tick rate, which could vary depending on refresh rate, throttling, etc meaning things won't be saved reliably

File: birb_registry/src/lib.rs
Comment (jw2476) [inline birb_registry/src/lib.rs:57]: LGTM

**Review by jw2476 — CHANGES_REQUESTED** [review]
I think we should change the tick-based saving and loading to just a save/load function that can be called by another module, otherwise the user has no control how the registry is saved. Also suggested a couple minor code style tweaks

**Comment by jw2476** [comment]
Thanks for renaming, the CI isn't going to work for a while since the core engine uses an unstable feature and the GH action only does stable rust afaik, from what I hear the feature will be stabilized on December 28th 🤞

File: birb_registry/src/lib.rs
Comment (camsoftworks2018) [reply birb_registry/src/lib.rs]: Yeah i realised this, adding in autosave functionality will be good for clock module

File: birb_registry/src/lib.rs
Comment (camsoftworks2018) [reply birb_registry/src/lib.rs]: One big unwrap chain, done

**Comment by Alex-Programs** [comment]
Agreed on giving users control over saving. We should probably move it to be run outside of the core engine loop at some point - just dispatch to another thread with a clone of the data and have some handling for a new save() when one is partway through - IO has high latency and we don't want to be slowing things down.

May also be worth looking into progressive/diff based saving depending on how large the registry is going to get. This could be implemented as simply as giving each key its own file and letting the FS handle it.

**Comment by jw2476** [comment]
The modules system is already multithreaded so I doubt moving it to a separate thread would do much. For now lets just get this merged and we can continue to iterate with improvements such as #9 after

**Comment by Alex-Programs** [comment]
Agreed, let's solve perf issues as they arise.

**Review by Alex-Programs — APPROVED** [review]
LGTM!

**Review by jw2476 — APPROVED** [review]
Looks good to me, nice work.

**Pull request merged**
